Framework/Crystal/src/RotateSampleShape.cpp

```diff
@@ -0,0 +1,179 @@
+    std::ostringstream propName;
+    propName << "Axis" << i;
+    declareProperty(std::make_unique<PropertyWithValue<std::string>>(propName.str(), "", Direction::Input),
+                    propName.str() + axisHelp);
+  }
+}
+
+/** Execute the algorithm.
+ */
+void RotateSampleShape::exec() {
+  Workspace_sptr ws = getProperty("Workspace");
+  auto ei = std::dynamic_pointer_cast<ExperimentInfo>(ws);
+
+  if (!ei) {
+    // We're dealing with an MD workspace which has multiple experiment infos
+    auto infos = std::dynamic_pointer_cast<MultipleExperimentInfos>(ws);
+    if (!infos) {
+      throw std::invalid_argument("Input workspace does not support RotateSampleShape");
+    }
+    if (infos->getNumExperimentInfo() < 1) {
+      ExperimentInfo_sptr info(new ExperimentInfo());
+      infos->addExperimentInfo(info);
+    }
+    ei = infos->getExperimentInfo(0);
+  }
+
+  std::string shapeXML;
+  bool isMeshShape = false;
+  if (!checkIsValidShape(ei, shapeXML, isMeshShape)) {
+    throw std::runtime_error("Input sample does not have a valid shape!");
+  }
+
+  // Create a goniometer with provided rotations
+  Goniometer gon;
+  prepareGoniometerAxes(gon, ei);
+  if (gon.getNumberAxes() == 0)
+    g_log.warning() << "Empty goniometer created; will always return an "
+                       "identity rotation matrix.\n";
+
+  const auto sampleShapeRotation = gon.getR();
+  if (sampleShapeRotation == Kernel::Matrix<double>(3, 3, true)) {
+    // If the resulting rotationMatrix is Identity, ignore the calculatrion
+    g_log.warning("Rotation matrix set via RotateSampleShape is an Identity matrix. Ignored rotating sample shape");
+    return;
+  }
+
+  const auto oldRotation = ei->run().getGoniometer().getR();
+  auto newSampleShapeRot = sampleShapeRotation * oldRotation;
+  if (isMeshShape) {
+    auto meshShape = std::dynamic_pointer_cast<MeshObject>(ei->sample().getShapePtr());
+    meshShape->rotate(newSampleShapeRot);
+  } else {
+    shapeXML = Geometry::ShapeFactory().addGoniometerTag(newSampleShapeRot, shapeXML);
+    Mantid::DataHandling::CreateSampleShape::setSampleShape(*ei, shapeXML, false);
+  }
+}
+
+bool RotateSampleShape::checkIsValidShape(const API::ExperimentInfo_sptr &ei, std::string &shapeXML,
+                                          bool &isMeshShape) {
+  if (ei->sample().hasShape()) {
+    const auto csgShape = std::dynamic_pointer_cast<CSGObject>(ei->sample().getShapePtr());
+    if (csgShape && csgShape->hasValidShape()) {
+      shapeXML = csgShape->getShapeXML();
+      if (!shapeXML.empty()) {
+        return true;
+      }
+    } else {
+      const auto meshShape = std::dynamic_pointer_cast<MeshObject>(ei->sample().getShapePtr());
+      if (meshShape && meshShape->hasValidShape()) {
+        isMeshShape = true;
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+void RotateSampleShape::prepareGoniometerAxes(Goniometer &gon, const API::ExperimentInfo_sptr &ei) {
```

Comment:
This appears to be the same as `SetGoniometer` code - I wonder if it's possible to refactor and put in a place it can be called from both algorithms? Like a goniometer helperfile?

Reply:
Although they appear to be the same code, there are some differences in terms of some additional validations, throw errors I have added in `RotateSampleShape` 's code and in the below line(`angle` was set as 0 in `SetGoniometer` alg)
```
gon.pushAxis(axisName, x, y, z, angle, ccw); 
```

If you think those validations are better to be included in SetGoniometer algorithm as well, then I can move to a utility class.

Reply:
OK, is there any reason why the validation has to be different in this case? If so then I'm OK to duplicate some code here - I definitely don't want to change the behaviour of SetGoniometer`) - also I see there's some difference s in

Reply:
1- Validation was done a little bit different to give more detailed information to the user such as
```
if (!Strings::convert(tokens[1], x))
        throw std::invalid_argument("Error converting x string '" + tokens[1] + "' to a number.");

```
vs
```
if (!Strings::convert(tokens[1], x))
          throw std::invalid_argument("Error converting string '" + tokens[1] + "' to a number.");
```

2- For `SetGoniometer` the below is a valid axis but not valid for `RotateSampleShape` as it needs the rotation angle as an input and hence needs a different validation.
`Axis0="Motor1,0,1,0,1"`


3- The name of the TimeSeriesProperty(`axisName`) added as a log value in `RotateSampleShape` below needs to differ from `SetGoniometer` as well to keep them separate when both algorithms are used for the same sample(which indeed can be done via helper class).
```
if (ei->mutableRun().hasProperty(axisName)) {
          ei->mutableRun().removeLogData(axisName);
```

Reply:
OK sounds good thanks!